rule_filter/lib/data/datasource_impl/rule_filter_datasource_impl.dart

```diff
@@ -0,0 +1,45 @@
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:rule_filter/data/datasource/rule_filter_datasource.dart';
+import 'package:rule_filter/data/network/rule_filter_api.dart';
+import 'package:rule_filter/rule_filter/tmail_rule.dart';
+
+class RuleFilterDataSourceImpl extends RuleFilterDataSource {
+  final RuleFilterAPI _ruleFilterAPI;
+
+  RuleFilterDataSourceImpl(this._ruleFilterAPI);
+
+  @override
+  Future<List<TMailRule>> getListTMailRule(AccountId accountId) {
+    return Future.sync(() async {
+      final listTMailRules = await _ruleFilterAPI.getListTMailRule(accountId);
+      return listTMailRules;
+    }).catchError((error) {
+      throw error;
+    });
+  }
+
+  @override
+  Future<void> updateListTMailRule(
+    List<TMailRule> listTMailRuleKeep,
+    List<TMailRule> listTMailRuleUpdate,
+    AccountId accountId,
+  ) {
+    final List<TMailRule> listTMailRule = List.empty(growable: true);
+    listTMailRule.addAll(listTMailRuleUpdate);
+    for (var itemTMailRuleKeep in listTMailRuleKeep) {
+      if (listTMailRuleUpdate
+          .where((itemTMailRuleUpdate) =>
+              itemTMailRuleKeep.id == itemTMailRuleUpdate.id)
```

Comment:
I checked, but all the time I did not see the rule id return when I get it. Please compare the object to determine it is duplicated or not.